README.md
# WiFi View
 
快速查看与管理在 Android 设备上连过的 WiFi 的密码
 
### 应用特性
- 支持 Android 4.0 +
- 支持显示中文名称的 WiFi
- 支持一键复制 SSID 与密码
- 支持备份与恢复
- 支持查看 WiFi 的源信息


## 应用截图
<figure class="pic">
    <a href="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/1.jpg"><img src="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/1.jpg" height="25%" width="25%" /></a>
    <a href="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/2.jpg"><img src="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/2.jpg" height="25%" width="25%" /></a>
    <a href="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/3.jpg"><img src="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/3.jpg" height="25%" width="25%" /></a>
    <a href="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/4.jpg"><img src="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/4.jpg" height="25%" width="25%" /></a>
    <a href="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/5.jpg"><img src="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/5.jpg" height="25%" width="25%" /></a>
    <a href="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/6.jpg"><img src="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/6.jpg" height="25%" width="25%" /></a>
    <a href="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/7.jpg"><img src="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/7.jpg" height="25%" width="25%" /></a>
    <a href="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/8.jpg"><img src="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/8.jpg" height="25%" width="25%" /></a>
    <a href="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/9.jpg"><img src="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/9.jpg" height="25%" width="25%" /></a>
    <a href="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/10.jpg"><img src="https://github.com/PCGJG/WiFi-View/blob/master/Screenshots/10.jpg" height="25%" width="25%" /></a>
</figure>

## 下载
- [GitHub](https://github.com/PCGJG/WiFi-View/blob/master/WiFi%20View.apk?raw=true)
- [酷安](https://www.coolapk.com/apk/com.zzz.wifiview)


## 隐私权限说明
### ACCESS_SUPERUSER:
使用 ROOT 权限读写 WiFi 密码配置文件。  
 
### ACCESS_WIFI_STATE:
扫描 WLAN ，用于置顶当前连接的 WiFi 。  
 
### ACCESS_FINE_LOCATION:
此项权限是 Android 官方要求的[[1]]。在 Android 8.0 以后，需要授予此权限才能使用 ACCESS_WIFI_STATE 权限。  
 
>软件扫描 WLAN 需要满足以下所有条件：  
> ① 授予软件精确位置权限  
> ② 设备已启用位置服务  
 
[[1]] https://developer.android.google.cn/guide/topics/connectivity/wifi-scan#wifi-scan-restrictions

[1]: <https://developer.android.google.cn/guide/topics/connectivity/wifi-scan#wifi-scan-restrictions>
 
 
## 注意

1. 手动删除的 WiFi 无法获取；

2. 部分系统可能会加密，导致软件不能正常工作；

3. 软件需要读取 /data 下的系统文件，必须授予 ROOT 权限。
  > Android 8.0 以下读取 /data/misc/wifi/wpa_supplicant.conf  
  > Android 8.0 及其以上读取 /data/misc/wifi/WifiConfigStore.xml

## License
```
Copyright (c) 2020 果酱稽

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```

*Copyright &#169; 2016-2020 果酱稽. All rights reserved.*


CHANGELOG.MD
# 历代更新日志


### 11 (39)
- 增加 权限弹窗限制
- 支持 浏览其他加密类型的 WiFi 信息
- 修复 计数菌动态更新
- 调整 通知提示内容

### 10 (38)
- 修复 当前连接的 WiFi 不能置顶的问题
- 调整 API 级别为 29 (Android 10)
- 调整 复制菜单
- 新增 隐私说明
- 新增 开放源代码

### 10 (37)
- 合并 复制菜单
- 优化 删除方法
- 优化 读取备份的信息传递
- 优化 无 ROOT 时的界面
- 修复 无法添加备份
- 修复 特殊备份名称引起的文件检索崩溃
- 修复 WiFi 条目不能删除的问题
- 修复 若干个 bug

### 9.0 (36)
- 删除 打开方式
- 调整 操作逻辑

### 8.1 (35)
- 删除 杂项
- 替换 刷新列表方式
- 精简 软件包

### 8.0 (34)
- 支持 当连接的 WiFi 在表头显示
- 支持 读取备份
- 支持 A-Z 排序
- 支持 使用 WiFi View 打开本地 WiFi 密码配置文件
- 支持 清空备份
- 修复 软件不自觉的创建备份文件夹
- 支持 修复 WiFi 配置文件出错导致的崩溃
- 优化 若干处若干次

### 7.1.2 (33)
- 恢复 删除 WiFi
- 新增 若干个功能

### 7.1 (32)
- 新增 备份与恢复
- 停用 WiFi 删除

### 7.0 (31)
- 新增 浏览源信息
- 支持 显示 WiFi 条目在源文件的位置
- 新增 WiFi 密码删除（不稳定)

### 6.0.1 (30)
- 紧急 版本回倒

### 6.0 (29)
- 支持 浏览源信息
- 优化 复制反馈

### 5.1.1 (28)
- 支持 Android 8.0
- 修正 正常的版本号
- 调整 targetSdkVersion
- 提高 运行处理的高效性
- 优化 代码的美观
- 检查 稳定性

### 5.0.2 (27)
- 支持 Android 8.0

### 5.0 (26)
- 修复版本号过低的问题

### 4.4.4 (25)
- 精简 软件
- 保留 核心功能

### 4.4 (24)
- 删除 无效代码
- 压制 图标
- 修改 蓝色主题

### 4.3 (23)
- 调整 计数菌不可点击
- 修复 看不懂英文的问题

### 4.2 (22)
- 支持 显示 WiFi 数量在 menu 里
- 支持 自定义文件目录
- 支持 导出分享

### 4.1 (21)
- 修复 跳转方式
- 精简 继续精简

### 4.0 (20)
- 新增 自定义备份名称

### 3.2 (19)
- 新增 菜单
- 新增 跳转到 WiFi 设置
- 新增 备份
- 修复 菜单问题
- 修复 时长问题

### 3.0 (18)
- 新增 备份
- 修复 菜单问题

### 2.9 (17)
- 新增 菜单
- 新增 跳转到WiFi设置

### 2.1 (16)
- 新增 计数菌 (=・ω・=)

### 2.6 (15)
- 新增 报错提示 （不要再问我为啥是空白界面了😅）

### 2.8 (14)
- 修改了菜单（样式如介绍图）

### 2.7 (13)
- 紧急修复

### 2.6 (12)
- 更换了新图标 192*192 px
- 修复了在部分设备上无法获取 root 的问题

### 2.5 (11)
- 移除了 values
- 移除了 debug 调试

### 2.4 (10)
- 继续精简，体积减小到 15 KB
- 删掉了无用的代码

### 2.3 (9)
- 调整 点击 item 为直接复制密码，长按打开复制菜单
- 除去 WiFi Setting

### 2.2 (8)
- 调整 点击 item 为直接复制密码，长按打开复制菜单
- 精简 代码
- 调整 item 高度
- 更换 Icon

### 2.1 (7)
- 精简

### 2.0 (6)
- 精简

### 1.1.1.1.1 (5)
- 精简 软件体积
- 除去 无用功能
- 支持 密码显示

### 1.1.1.1 (4)
- 新增 快速跳转到 WiFi 设置
- 优化 界面布局

### 1.1.1 (3)
- 新增 提示界面
- 删除 其他尺寸的图标

### 1.1 (2)
- 新增 WiFi 密码查看
- 新增 WiFi 开关
- 新增 退出确认

### 1.0 (1)
Hello World!
 
## - END -
 


app/src/java/com/zzz/wifiview/ViewActivity.java
package com.zzz.wifiview;

import android.app.ActionBar;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.PopupMenu;
import android.widget.Toast;
import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Map;

public class ViewActivity extends Activity {
    ArrayList<Map<String, String>> mainList;
    PopupMenu popup;
    Context context = this;
    String backupPath; // 备份目录 (/AppData/files/Backup/AutoBackup)
    String sPath; // 源文件
    boolean isRead = false; // 读取备份模式
    boolean isMore = false; // 显示更多 WiFi
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.list);
        
        // 缓存目录(/AppData/cache/bk)
        //bkPath = this.getExternalCacheDir().getPath() + "/bk";
        
        // 备份目录 (/AppData/files/Backup/LastBackup)
        backupPath = this.getExternalFilesDir("Backup").getPath() + "/AutoBackup";
        
        try {
            Intent intent = getIntent();
            Bundle bundle = intent.getExtras();
            isRead = bundle.getBoolean("read");
            isMore = bundle.getBoolean("more");
        } catch (Exception e) {
            e.printStackTrace();
            //Toast.makeText(this, "VAoC: " + e.getMessage(), Toast.LENGTH_LONG).show();
        }
        
        start();
    }
    
    
    private void start() {
        Intent intent = getIntent();
        Bundle bundle = intent.getExtras();
        
        /* 判断启动来源 */
        if (isRead) { // 读取备份模式
            setTitle(bundle.getString("name"));
            sPath = bundle.getString("path");
            mainList = get(sPath, 1);
        } else if (isMore) { // 显示更多 WiFi
            setTitle("更多 WiFi");
            sPath = bundle.getString("path");
            mainList = get(sPath, 2);
        } else { // 读取系统文件
            if (android.os.Build.VERSION.SDK_INT >= 26) {
                sPath = "/data/misc/wifi/WifiConfigStore.xml";
                mainList = get(sPath, 1);
            } else {
                sPath = "/data/misc/wifi/wpa_supplicant.conf";
                mainList = get(sPath, 1);
            }
            
        }
        
        
        if (mainList == null) {
            Toast.makeText(ViewActivity.this, "获取列表失败", Toast.LENGTH_LONG).show();
        } else {
            if (mainList.size() == 0) {
                Toast.makeText(ViewActivity.this, "列表为空", Toast.LENGTH_LONG).show();
            } else {
                final ListView lv = findViewById(R.id.lv);
                lv.setAdapter(new WiFiAdapter(this, mainList));
                lv.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                    @Override
                    public void onItemClick(AdapterView<?> parent, View view, final int position, long id) {
                        popup = new PopupMenu(ViewActivity.this, view);
                        
                        if (isMore) {
                            getMenuInflater().inflate(R.menu.more,popup.getMenu());
                        } else {
                            getMenuInflater().inflate(R.menu.copy,popup.getMenu());
                        }
                        
                        popup.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
                            @Override
                            public boolean onMenuItemClick(MenuItem item) {
                                final ClipboardManager cm = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
                                switch(item.getItemId()) {
                                    case R.id.menu_ssid:
                                        cm.setPrimaryClip(ClipData.newPlainText(null, mainList.get(position).get("ssid")));
                                        Toast.makeText(ViewActivity.this, "SSID已复制", Toast.LENGTH_SHORT).show();
                                        break;
                                    case R.id.menu_password:
                                        cm.setPrimaryClip(ClipData.newPlainText(null, mainList.get(position).get("psk")));
                                        Toast.makeText(ViewActivity.this, "密码已复制", Toast.LENGTH_SHORT).show();
                                        break;
                                    case R.id.menu_all:
                                        Map<String, String> s = mainList.get(position);
                                        cm.setPrimaryClip(ClipData.newPlainText(null, "SSID: " + s.get("ssid") + "\n" + "密码: " + s.get("psk")));
                                        Toast.makeText(ViewActivity.this, "SSID和密码都已复制", Toast.LENGTH_SHORT).show();
                                        break;
                                    case R.id.menu_view:
                                        AlertDialog.Builder TextDialog = new AlertDialog.Builder(ViewActivity.this);
                                        TextDialog.setTitle("源信息浏览").setMessage(mainList.get(position).get("view")).setPositiveButton("关闭",null).setNeutralButton("复制", new DialogInterface.OnClickListener(){@Override public void onClick(DialogInterface dialog, int which) {
                                            cm.setPrimaryClip(ClipData.newPlainText(null, mainList.get(position).get("view")));
                                            }}).show();
                                        break;
                                    case R.id.menu_delete:
                                        //run("cp -f " + wifiPath + " " + backupPath);
                                        delete(mainList.get(position).get("view"));
                                        cmd("cp -f " + backupPath + " " + sPath);
                                        if (!isRead) cmd("chmod 660 " + sPath);
                                        Toast.makeText(ViewActivity.this, "已删除", Toast.LENGTH_SHORT).show();
                                        
                                        mainList.clear();
                                        start();
                                        
                                        break;
                                    default:
                                        return false;
                                }
                                return true;
                            }
                        });
                        popup.show();
                    }
                });
            }
        }
    
    }
    
    
    public ArrayList<Map<String, String>> get(String path, int i) {
        try {
            ReadFile file = new ReadFile(path,i);
            return file.getList(this.context);
        } catch (Exception e) {
            Toast.makeText(this, "VAget: " + e.getMessage(), Toast.LENGTH_LONG).show();
            return null;
        }
    }
    
    
    @Override
    protected void onStart() {
        super.onStart();
        if (isMore) {
            ActionBar actionBar = this.getActionBar();
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
    
    }
    
    
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        menu.add(0,0,0,"刷新列表");
        menu.add(0,1,0,"显示更多 WiFi");
        menu.add(0,2,0,"打开 WiFi 设置");
        menu.add(0,3,0,"备份与恢复");
        menu.add(0,4,0,"关于");
        menu.add(0,5,0,"获取列表出错");
        menu.getItem(5).setEnabled(false);
        if (mainList != null) menu.getItem(5).setTitle("共 " + mainList.size() + " 条WiFi");
        if (isRead || isMore) {
            menu.getItem(2).setVisible(false);
            menu.getItem(3).setVisible(false);
            menu.getItem(4).setVisible(false);
        }
        if (isRead) {
            menu.getItem(0).setVisible(false);
            menu.add(0,6,0,"返回");
        }
        if (isMore) menu.getItem(1).setVisible(false);
        return true;
    }
    
    
    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);
        if (mainList != null) menu.getItem(5).setTitle("共 " + mainList.size() + " 条WiFi");
        return super.onPrepareOptionsMenu(menu);
    }
    
    
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case 0: // 刷新列表
                mainList.clear();
                start();
                Toast.makeText(this, "刷新成功", Toast.LENGTH_SHORT).show();
                return true;
            case 1: // 显示更多 WiFi
                Intent intent = new Intent();
                intent.setClass(ViewActivity.this, ViewActivity.class);
                intent.putExtra("path", sPath);
                intent.putExtra("more", true);
                startActivity(intent);
                return true;
            case 2: // 打开 WiFi 设置
                startActivity(new Intent(Settings.ACTION_WIFI_SETTINGS));
                return true;
            case 3: // 备份与恢复
                startActivity(new Intent().setClassName("com.zzz.wifiview","com.zzz.wifiview.FileActivity"));
                return true;
            case 4: // 关于
                showAboutDialog();
                return true;
            case 6: // 返回
                finish();
                return true;
            case android.R.id.home: // 左上角返回
                finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
    
    
    private void delete(String ss) {
        String s = "";
        Process process = null;
        DataOutputStream os = null;
        BufferedReader br = null;
        OutputStreamWriter pw = null;
        try {
            process = Runtime.getRuntime().exec("su");
            os = new DataOutputStream(process.getOutputStream());
            os.writeBytes("cat " + sPath + "\n");
            os.writeBytes("exit\n");
            os.flush();
            
            br = new BufferedReader(new InputStreamReader(process.getInputStream()));
            String line = null;
            while ((line = br.readLine()) != null) {
                s += line.trim() + "\n";
            }
            pw = new OutputStreamWriter(new FileOutputStream(backupPath),"UTF-8");
            s = s.replace(ss,"");
            pw.write(s);
        } catch (IOException e) {
            e.printStackTrace();
            Toast.makeText(ViewActivity.this, "VAdelete: " + e, Toast.LENGTH_LONG).show();
        } finally {
            try {
                if (os != null) {
                    os.close();
                }
                if (br != null) {
                    br.close();
                }
                if (pw != null) {
                    pw.close();
                }
                process.destroy();
            } catch (Exception e) {
                e.printStackTrace();
                Toast.makeText(this, "VAdeleteF: " + e.getMessage(), Toast.LENGTH_LONG).show();
            }
        }
    }
    
    
    private void showAboutDialog() {
        final String[] items = {"版本: 11", "开发者: 果酱稽", "应用信息", "在应用市场中查看", "开放源代码", "Copyright © 2016 - 2020 JamG.\nAll rights reserved."};
        AlertDialog.Builder AboutDialog = new AlertDialog.Builder(this)
            .setTitle("WiFi View")
            .setIcon(com.zzz.wifiview.R.drawable.ic)
            .setItems(items, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    switch (which) {
                        case 0:
                            Toast.makeText(ViewActivity.this, "版本代号: 39", Toast.LENGTH_SHORT).show();
                            break;
                        case 1:
                            Toast.makeText(ViewActivity.this, "pcgjg1001@gmail.com", Toast.LENGTH_LONG).show();
                            break;
                        case 2:
                            Intent intent = new Intent();
                            intent.setAction(android.provider.Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                            intent.setData(Uri.parse("package:com.zzz.wifiview"));
                            startActivity(intent);
                            break;
                        case 3:
                            Uri uri1 = Uri.parse("https://www.coolapk.com/apk/com.zzz.wifiview");
                            Intent intent1 = new Intent(Intent.ACTION_VIEW, uri1);
                            startActivity(intent1);
                            break;
                        case 4:
                            Uri uri2 = Uri.parse("https://github.com/PCGJG/WiFi-View");
                            Intent intent2 = new Intent(Intent.ACTION_VIEW, uri2);
                            startActivity(intent2);
                            break;
                    }
                }
            });
        AboutDialog.create().show();
    }
    
    
    public void cmd(String command) {
        Process process = null;
        DataOutputStream os = null;
        try {
            process = Runtime.getRuntime().exec("su");
            os = new DataOutputStream(process.getOutputStream());
            os.writeBytes(command + "\n");
            os.writeBytes("exit\n");
            os.flush();
            process.waitFor();
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(this, "VAcmd: " + e.getMessage(), Toast.LENGTH_LONG).show();
        } finally {
            try {
                if (os != null) {
                    os.close();
                }
                process.destroy();
            } catch (Exception e) {
                e.printStackTrace();
                Toast.makeText(this, "VAcmdF: " + e.getMessage(), Toast.LENGTH_LONG).show();
            }
        }
    }
    
    
}
